### DISCUSSION TITLE
How can I use PEP 681 to actually provide "dataclass-like semantics" to custom classes?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [microsoft/pyright#7178] [Q&A] [answered]
I am trying to find out how I can use the information in [PEP 681](https://peps.python.org/pep-0681/) to write a base class decorated with `dataclass_transform` that enables derived classes with the exact same semantics as Python dataclasses. Specifically, I am interested in using field specifiers, but I came across the following behavior of Pyright 1.1.349:

```python
import typing
from typing import Any, cast
from dataclasses import dataclass, field


# A plain old dataclass.
# Note that `field` returns a `Field` object, but `x` is declared as `int`.
@dataclass
class MyDataclass:
    x: int = field(default=1)  # no error here


# The descriptor class from https://github.com/microsoft/pyright/issues/3245
# Note that it is declared to get and set `int` values.
class MyDescriptor:
    def __get__(self, __obj: object | None, __owner: Any) -> "int | MyDescriptor":
        del __owner
        if __obj is None:
            return self
        return cast(Any, __obj)._x

    def __set__(self, __obj: object | None, __value: int) -> None:
        if __obj is not None:
            cast(Any, __obj)._x = __value


# A class that should give "dataclass-like semantics" to its subclasses,
# from https://peps.python.org/pep-0681/
@typing.dataclass_transform(field_specifiers=(MyDescriptor,))
class ModelBase: ...


class Foo(ModelBase):
    x: MyDescriptor = MyDescriptor()  # no error here
    y: int = MyDescriptor()  # error here:
    # Expression of type "MyDescriptor" cannot be assigned to declared type "int" Pyright (reportAssignmentType)
    #   "MyDescriptor" is incompatible with "int"
```

Note that with the `dataclass` decorator, you can directly specify the type of the attribute managed by the `field` descriptor as `int`, while for my custom class this does not seem to be the case. Is it possible to get the exact same behavior for my custom class, and if so, would I have to make changes to `MyDescriptor`, `ModelBase` or something else?

**Comment by NiklasKappel** [comment, marked as answer]
I learned from attrs that the trick is to lie about the type of the descriptor: https://github.com/python-attrs/attrs/blob/75d54c219771b30570b29aee1293d28dc803236e/src/attr/__init__.pyi#L149

**Reply by erictraut** [reply]
Yes, that's the technique that `dataclass.field` uses as well, and it's the approach that static type checkers assume for field specifier classes.

**Reply by NiklasKappel** [reply]
Thanks for confirming!

**Comment by NiklasKappel** [comment]
> Yes, that's the technique that `dataclass.field` uses as well, and it's the approach that static type checkers assume for field specifier classes.

I have a follow-up question: I understand that we can easily lie about the return type of a factory function like `dataclass.field` or `attrs.attrib`. But can we avoid the use of factory functions and directly set types for the constructors of field specifiers, like `MyDescriptor` in `y: int = MyDescriptor()`? I know that `__init__` does not actually return anything, so maybe one can type `__new__`? If it was possible to type a descriptor constructor, would it also be possible to propagate the type hints to constructors of classes that inherit from the original descriptor?

**Reply by erictraut** [reply]
Sorry, I don't understand what you mean by "directly set types for the constructors of field specifiers". Perhaps you could provide an example in code?

**Reply by NiklasKappel** [reply]
Forgive my ramblings and consider:
```python
from typing import TypeVar, dataclass_transform


class DescriptorBase:
    def __init__(self, value):
        self.value = value


class DescriptorChildA(DescriptorBase): ...


class DescriptorChildB(DescriptorBase): ...


_T = TypeVar("_T")


def descriptor_factory(value: _T) -> _T:
    return DescriptorBase(value)  # type: ignore  # no problem if in stub file


@dataclass_transform(field_specifiers=(DescriptorBase, DescriptorChildA, DescriptorChildB))
class ModelBase: ...


class Model(ModelBase):
    # None of these check out
    x: int = DescriptorBase(1)
    y: int = DescriptorChildA(2)
    z: int = DescriptorChildB(3)

    # This one is fine
    w: int = descriptor_factory(4)

```
If I type hint the `descriptor_factory` function, pyright understands that the type of the `w` field in `Model` should be `int`, just like when I use the `dataclasses.field` descriptor. In my specific use case, I have a `DescriptorBase` class and multiple child classes that inherit from it. The child classes all use the exact same `__init__`, they just add various unrelated functionality.

I would like to get dataclass-like semantics for all my descriptor classes. I could now write a factory function for each child class and and declare return types for the factory functions in a stub file, just like `attrs` does. However, because all child classes use the exact same `__init__`, the type hints for their factory functions would all be identical, which yields a lot of boilerplate code.

My question is: Is it possible to make the types check out, without using factory functions, just by using the descriptor constructors directly like for `x`, `y`, and `z`? If so, what is the minimal amount of stub file boilerplate I need to write? Is it somehow possible to just type `__init__` of `DescriptorBase`, which all descriptor classes use anyways?

**Reply by erictraut** [reply]
Ah, I see what you mean.

> Is it possible to make the types check out, without using factory functions...

No, that's not possible because that's not how `dataclass` and its field specifiers were designed, and `dataclass_transform` simply adopts the design of `dataclass`. 

You could minimize boilerplate by creating a simple wrapper that transforms a descriptor class into a "field specifier factory".

```python
def create_field_factory[**P](cls: Callable[P, Any]) -> Callable[P, Any]:
    return cls

descriptor_base = create_field_factory(DescriptorBase)
descriptor_child_A = create_field_factory(DescriptorChildA)
descriptor_child_B = create_field_factory(DescriptorChildB)

...

class Model(ModelBase):
    x: int = descriptor_base(1)
    y: int = descriptor_child_A(2)
    z: int = descriptor_child_B(3)
```

**Reply by athyuttamre** [reply]
@NiklasKappel I'm also running into the same issue — wanting to give dataclass semantics without creating a wrapper function. Did you find a workaround by any chance?
